src/interrupt/kvm/msi_irq.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright (C) 2019 Alibaba Cloud. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Helper utilities for handling MSI interrupts.
+
+use super::*;
+use kvm_bindings::{kvm_irq_routing_entry, KVM_IRQ_ROUTING_MSI};
+
+pub(super) struct MsiConfig {
+    pub(super) irqfd: EventFd,
+    pub(super) config: Mutex<MsiIrqSourceConfig>,
+}
+
+impl MsiConfig {
+    pub(super) fn new() -> Self {
+        MsiConfig {
+            irqfd: EventFd::new(0).unwrap(),
+            config: Mutex::new(Default::default()),
+        }
+    }
+}
+
+pub(super) fn new_msi_routing_entry(
+    gsi: InterruptIndex,
+    msicfg: &MsiIrqSourceConfig,
+) -> kvm_irq_routing_entry {
+    let mut entry = kvm_irq_routing_entry {
+        gsi,
+        type_: KVM_IRQ_ROUTING_MSI,
+        flags: 0,
+        ..Default::default()
+    };
+    unsafe {
+        entry.u.msi.address_hi = msicfg.high_addr;
+        entry.u.msi.address_lo = msicfg.low_addr;
+        entry.u.msi.data = msicfg.data;
+    }
+    entry
+}
+
+#[allow(irrefutable_let_patterns)]
+pub(super) fn create_msi_routing_entries(
```

Comment:
Method `create_msi_routing_entries` cares MSI only, but it takes generic type "InterruptSourceConfig" as parameter. It would be better to align with method `new_msi_routing_entry` above by:
- Changing parameter type of `configs` to `MsiIrqSourceConfig`. It requires some type adaptation work in `pci_msi_irq.rs`. 
- Changing the name from `create_msi_routing_entries` to `new_msi_routing_entries`, or changing the name `new_msi_routing_entry` to `create_msi_routing_entry`.

Reply:
Function create_msi_routing_entries() will be called by InterruptSourceGroup::enabled(), it's more convenient to use InterruptSourceConfig instead of MsiIrqSourceConfig. Actually the main work of create_msi_routing_entries() is to convert parameter from InterruptSourceConfig to MsiIrqSourceConfig.